### PR TITLE
=htc #19827 avoid double-push when request timeout triggers

### LINKED ISSUE #19827: Akka-http request timeout throws an exception
```
016-02-18T19:50:14,103Z [sync-server-akka.actor.default-dispatcher-8] ERROR a.a.RepointableActorRef:71 - Error in stage [akka.http.impl.engine.server.HttpServerBluePrint$RequestTimeoutSupport@31ef980d]: requirement failed: Cannot push port (responseOut) twice
java.lang.IllegalArgumentException: requirement failed: Cannot push port (responseOut) twice
	at scala.Predef$.require(Predef.scala:219) ~[scala-library-2.11.7.jar:na]
	at akka.stream.stage.GraphStageLogic.push(GraphStage.scala:439) ~[akka-stream_2.11-2.4.2.jar:na]
	at akka.http.impl.engine.server.HttpServerBluePrint$RequestTimeoutSupport$$anon$6$$anon$9.onPush(HttpServerBluePrint.scala:292) ~[akka-http-core_2.11-2.4.2.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:582) ~[akka-stream_2.11-2.4.2.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:593) ~[akka-stream_2.11-2.4.2.jar:na]
```

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/akka/akka/issues/19827

The issue was that when the timeout happened, it called `emit(responseOut, response._2, () ⇒ complete(responseOut))` ([line 280](https://github.com/akka/akka/pull/19871/files#diff-623ef36775342f2db6b86855f7ffea64R280)) so the `responseOut` may already be closed by the timeout once we get a chance to attempt pushing the real response. "Sadly it didn't make it in time – we sent a response and closed the port already."

I'm not fully sure if we want to log the truncation error though?

### CODE REVIEW

**Comment by johanandren** [comment]
LGTM

I'd say we don't want the truncation error, but maybe that is hard to change only for this specific case.

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala [hunk 1]
```diff
@@ -289,7 +289,10 @@ private[http] object HttpServerBluePrint {
         def onPush(): Unit = {
           openTimeouts.head.clear()
           openTimeouts = openTimeouts.tail
-          push(responseOut, grab(responseIn))
+
+          // it may have been completed by a timeout already
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: the question is, why do we get an onPush after a timeout? Isn't the correct solution is to cancel this port on timeout so it doesn't get pushed (we ignore the push anyway)?

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: cancel this port is happening actually (thus the `isAvailable` == false). 
I did not try `completeStage()` instead `complete(responseOut)` actually... need to think if it changes semantics or not hm

Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: If it (`responseIn`) has been cancelled then this fix makes no sense as onPush cannot arrive for a cancelled port.

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: only out was cancelled, completeStage effectively cancels both which actually may be what we need here - testing currently.

Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: hm, I see it is just isAvailable and not isClosed. But then responseOut is not closed yet. So maybe I am just confused. Anyway, the fix is not self-evident ("it's your fault!" ;))

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: ;) Tracked the execution in more detail and pushed new rev, I'll explain as top level comment

**Comment by ktoso** [comment]
Added high level directives for request timeouts as well as documentation for all timeouts.
Fix is completing the stage, as discussed on gitter.

**Comment by ktoso** [comment]
Resolves https://github.com/akka/akka/issues/19896

File: akka-docs/rst/scala/http/common/timeouts.rst [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+Server timeouts
+---------------
+
+.. _request-timeout:
+
+Request timeout
+^^^^^^^^^^^^^^^
+
+Request timeouts are a mechanism that limits the maximum time it may take to produce an ``HttpResponse`` from a route.
+If that deadline is not meat the server will automatically inject a Service Unavailable HTTP response and close the connection
```
Comment (johanandren) [inline akka-docs/rst/scala/http/common/timeouts.rst]: deadline is **met**

File: akka-docs/rst/scala/http/routing-dsl/directives/alphabetically.rst [hunk 3]
```diff
@@ -210,10 +210,14 @@ Directive                                   Description
 :ref:`-tprovide-`                           Injects a given tuple of values into a directive
 :ref:`-uploadedFile-`                       Streams one uploaded file from a multipart request to a file on disk
 :ref:`-validate-`                           Checks a given condition before running its inner route
+:ref:`-withoutRequestTimeout-`              Disables :ref:`request timeouts <request-timeout>` for a given route.
```
Comment (johanandren) [inline akka-docs/rst/scala/http/routing-dsl/directives/alphabetically.rst]: cool!

File: akka-docs/rst/project/sponsors.rst [hunk 4]
```diff
@@ -4,7 +4,7 @@ Sponsors
 ========
 
 Lightbend
---------
+---------
 
 Lightbend is the company behind the Akka Project, Scala Programming Language,
 Play Web Framework, Scala IDE, Simple Build Tool and many other open source
```
Comment (patriknw) [inline akka-docs/rst/project/sponsors.rst]: s/Simple Build Tool/sbt/

Comment (ktoso) [inline akka-docs/rst/project/sponsors.rst]: lol, good catch :)

File: akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        path("timeout") {
+          withRequestTimeout(3.seconds) {
+            val response: Future[String] = slowFuture() // very slow
+            complete(response)
+          }
+        }
+      //#
+    }
+    "without timeout" in {
+      pending // compile only spec since requires actuall Http server to be run
```
Comment (patriknw) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: should we use `pending` for this? we will never "fix" it. perhaps a `if (false)` instead?

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import org.scalatest.{ BeforeAndAfterAll, Matchers, WordSpecLike }
+import scala.concurrent.duration._
+import scala.concurrent.Await
+
+/** INTERNAL API - not (yet?) ready for public consuption */
+private[akka] trait IntegrationRoutingSpec extends WordSpecLike with Matchers with BeforeAndAfterAll
+  with Directives with RequestBuilding
+  with ScalaFutures with IntegrationPatience {
+
+  implicit val system = ActorSystem(getClass.getSimpleName + "-System")
```
Comment (patriknw) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala]: is this name correct? we do crazy things in AkkaSpec to get at the class name from a constructor, but perhaps it is different for a `val`?

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  import system._
```
Comment (patriknw) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala]: importing everything from system, really?

**Comment by patriknw** [comment]
LGTM I think

File: akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala (hunk 5 above)
Comment (johanandren) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: in some specs we have there is a `compileTimeOnly(block: => Any)` that is pretty nice

Comment (patriknw) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: nice

Comment (ktoso) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: Yeah, let's use that everywhere :)

Comment (ktoso) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: I'll make it accessible in all doc tests

File: akka-docs/rst/scala/http/common/timeouts.rst (hunk 2 above)
Comment (ktoso) [inline akka-docs/rst/scala/http/common/timeouts.rst]: :meat_on_bone:

Comment (johanandren) [inline akka-docs/rst/scala/http/common/timeouts.rst]: if (deadline != :meat_on_bone:) inject service unavailable response

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala (hunk 6 above)
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala]: Likely not, I'll copy the way AkkaSpec does

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala (hunk 7 above)
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/IntegrationRoutingSpec.scala]: fixed, just dispatcher

**Comment by ktoso** [comment]
Addressed comments

**Comment by ktoso** [comment]
Let's merge after validation

**Comment by johanandren** [comment]
LGTM!

**Pull request merged**